packages/cozy-ci/push_artifact.js

```diff
@@ -0,0 +1,157 @@
+  const options = {
+    gzip: true,
+    cwd: folderName,
+    file: folderName + '/' + archiveFileName
+  }
+  try {
+    await tar.c(options, fileList)
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(
+      `↳ ❌ Unable to generate app archive. Is tar installed as a dependency ? Error : ${
+        error.message
+      }`
+    )
+    throw new Error('Unable to generate archive')
+  }
+}
+
+const pushArtifact = async (fileName, parentFolder, options) => {
+  const { appSlug, appVersion, buildCommit } = options
+  // eslint-disable-next-line no-console
+  console.warn(`↳ ℹ️  Sending artifact to downcloud`)
+  const folder = `${appSlug}/mobile/${appVersion}${
+    buildCommit ? `-${buildCommit}` : ''
+  }/`
+  try {
+    await launchCmd('ssh', [
+      '-o',
+      'StrictHostKeyChecking=no',
+      'upload@downcloud.cozycloud.cc',
+      `mkdir -p ${DOWNCLOUD_UPLOAD_DIR}${folder}`
+    ])
+  } catch (e) {
+    throw new Error(
+      `Unable to create target directory ${folder} on downcloud server : ${
+        e.message
+      }`
+    )
+  }
+
+  await launchCmd(
+    'rsync',
+    [
+      // to remove host validation question on CI
+      '-e',
+      'ssh -o StrictHostKeyChecking=no',
+      '-a',
+      fileName,
+      `upload@${DOWNCLOUD_URL}:${DOWNCLOUD_UPLOAD_DIR}${folder}`
+    ],
+    { cwd: parentFolder }
+  )
+
+  // eslint-disable-next-line no-console
+  console.warn(`↳ ℹ️  Upload to downcloud complete.`)
+  return {
+    ...options,
+    appBuildUrl: `https://${DOWNCLOUD_URL}/upload/${folder}${fileName}`
+  }
+}
+
+const getUploadTarget = async (uploadTarget, appName) => {
+  if (!fs.existsSync(uploadTarget)) {
+    throw new Error(`❌ ${uploadTarget} does not exist.`)
+  } else if (fs.lstatSync(uploadTarget).isDirectory()) {
+    const archiveFileName = `${appName.toLowerCase()}.tar.gz`
+    await createArchive(uploadTarget, archiveFileName)
+
+    return [uploadTarget, archiveFileName]
+  } else {
+    const uploadDir = path.dirname(uploadTarget)
+    const uploadFile = path.basename(uploadTarget)
+
+    return [uploadDir, uploadFile]
+  }
+}
+
+const run = async () => {
+  try {
+    const uploadTarget = process.argv[2]
+    const packageFile = process.argv[3] ? process.argv[3] : '../../package.json'
+    let { version, name } = require(packageFile)
+
+    const [uploadDir, uploadFile] = await getUploadTarget(uploadTarget, name)
+
+    const { appBuildUrl } = await pushArtifact(uploadFile, uploadDir, {
+      appSlug: name,
+      appVersion: version,
+      buildCommit: process.env.TRAVIS_COMMIT
+    })
+    // eslint-disable-next-line no-console
+    console.warn(`↳ ✅ Upload complete, artifact available at ${appBuildUrl}`)
+    // eslint-disable-next-line no-console
+    console.log(appBuildUrl)
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(`↳ ❌  Error while uploading: ${error.message}`)
+    process.exit(1)
+  }
+}
+
+run()
```

Comment:
Some parts looks a lot like things we have in `cozy-app-publish`. Do you think there is something to do to be more DRY? Would it be worth it?

Reply:
Yes but this script was created by @y-lohse I just copy it and adapt some lines